sway-lib-std/src/ecr.sw

```diff
@@ -1,31 +1,44 @@
 library ecr;
 
-use ::b512::B512;
 use ::address::Address;
+use ::b512::B512;
+use ::constants::ZERO;
+use ::hash::{HashMethod, hash_pair};
+use ::result::Result;
+
+pub enum EcRecoverError {
+    UnrecoverableAddress: (),
+    UnrecoverablePublicKey: (),
+}
 
 /// Recover the public key derived from the private key used to sign a message
-pub fn ec_recover(signature: B512, msg_hash: b256) -> B512 {
+pub fn ec_recover(signature: B512, msg_hash: b256) -> Result<B512, EcRecoverError> {
     let public_key = ~B512::new();
 
     asm(buffer: public_key.bytes, sig: signature.bytes, hash: msg_hash) {
         ecr buffer sig hash;
     };
-
-    public_key
+    /// if recovered key is empty
+    if (public_key.bytes)[0] == ZERO && (public_key.bytes)[1] == ZERO {
```

Comment:
This isn't how `ecr` works. From the specs:

> If the signature cannot be verified, `MEM[$rA, 64]` is set to `0` and `$err` is set to `1`, otherwise `$err` is cleared.

Note that `$err` being set to `1` with default flags with panic. In other words, you won't even be able to check the value of the memory range since the VM will panic before that!

The way to do this is to have two ec_recover functions:
1. One that panics immediately on error.
2. One that returns a result. This requires first disabling panic-on-error, then re-enabling it after the `ecr` instruction.

Reply:
@adlerjohn which flag is used to disable panic-on-error, is it:
<img width="648" alt="image" src="https://user-images.githubusercontent.com/17053076/162111990-1a1591bf-7f17-4dc8-8736-b8882f9c04ea.png">
?
If so, I'm not sure if using the `flag` opcode just toggles the flag on and off ? ie:
```
asm(buffer: public_key.bytes, sig: signature.bytes, hash: msg_hash) {
        flag i2;// disable panic-on-error
        ecr buffer sig hash;
        flag i2;// reenable panic-on-error
    };
```

Note that I'm not sure an immediate will work here,  may need to initialize a register with the value.

Reply:
Oh wait, I forgot how the thing I designed works.

The [header of ALU instructions](https://github.com/FuelLabs/fuel-specs/blob/c8e31b02ae8ed14156dc54a6a28b691bbfcb16e2/specs/vm/opcodes.md#arithmeticlogic-alu-opcodes) says
> All these opcodes advance the program counter $pc by 4 after performing their operation.
> If the [F_UNSAFEMATH](https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/main.md#flags) flag is unset, an operation that would have set $err to true is instead a panic.
> If the [F_WRAPPING](https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/main.md#flags) flag is unset, an operation that would have set $of to a non-zero value is instead a panic.

In other words, for regular execution of _other_ instructions (including ecrecover), `$err` being set to `1` is _not_ an automatic panic, my bad.

So the way you'd use it is rather than comparing that the 64 bytes are zero, just check the `$err` register. Much simpler than what I proposed above 😂

Reply:
Ok, this is very helpful! Thanks

Reply:
Note that any other instruction (I think) will reset the `$err` register value, so you either want to compare it immediately in asm (not the best), or `move` it into another register. Can asm blocks return a tuple?

Reply:
I'm wondering if I can do something roughly like this, as the asm block was not explicitly returning anything before:
```
pub fn ec_recover(signature: B512, msg_hash: b256) -> Result<B512, EcRecoverError> {
    let public_key = ~B512::new();

    let was_error = asm(buffer: public_key.bytes, sig: signature.bytes, hash: msg_hash) {
        ecr buffer sig hash;
        err
    };

    // check the $err register to see if the `ecr` opcode succeeded
    if was_error == 1 {
        Result::Err(EcRecoverError::UnrecoverablePublicKey)
    } else {
        Result::Ok(public_key)
    }
}
```

Reply:
Looks right

Reply:
OK, tests passed locally, and I added a couple test cases for failed recovery.
cc @adlerjohn